xls/dslx/interpreter_main.cc

```diff
@@ -169,16 +173,31 @@ absl::StatusOr<TestResult> RealMain(
   XLS_ASSIGN_OR_RETURN(std::string module_name, PathToName(entry_module_path));
 
   std::unique_ptr<AbstractRunComparator> run_comparator;
-  switch (compare_flag) {
-    case CompareFlag::kNone:
-      break;
-    case CompareFlag::kJit:
-      run_comparator = std::make_unique<RunComparator>(CompareMode::kJit);
-      break;
-    case CompareFlag::kInterpreter:
-      run_comparator =
+  // When comparison is on, use the same mode (JIT/interpreter) as the
+  // comparator. Otherwise, use the same mode as the evaluator flag. Using the
+  // IR interpreter requires --run_quickcheck_when_interpreting.
+  std::unique_ptr<AbstractRunComparator> quickcheck_runner;
+  if (compare_flag == CompareFlag::kNone) {
+    if (evaluator == EvaluatorType::kIrJit) {
+      quickcheck_runner = std::make_unique<RunComparator>(CompareMode::kJit);
+    } else if (evaluator == EvaluatorType::kIrInterpreter) {
+      if (absl::GetFlag(FLAGS_run_quickcheck_when_interpreting)) {
+        quickcheck_runner =
+            std::make_unique<RunComparator>(CompareMode::kInterpreter);
+      }
+    } else {
+      // Quickcheck is never run with the DSLX interpreter.
+    }
+  } else if (compare_flag == CompareFlag::kJit) {
```

Comment:
should we print or error out if `evaluator` is different?

Reply:
Good idea. The thing to check is that the evaluator is the DSLX interpreter in this case. The comparison mode checks the result of the DSLX interpreter against the specified evaluator.